ArduPlane/mode_loiter.cpp

```diff
@@ -42,28 +42,71 @@ bool ModeLoiter::isHeadingLinedUp(const Location loiterCenterLoc, const Location
     // Return true if current heading is aligned to vector to targetLoc.
     // Tolerance is initially 10 degrees and grows at 10 degrees for each loiter circle completed.
 
-    if (loiterCenterLoc.get_distance(targetLoc) < 1.05f * fabsf(plane.loiter.radius)) {
-        /* Whenever next waypoint is within the loiter radius plus 5%,
-           maintaining loiter would prevent us from ever pointing toward the next waypoint.
-           Hence break out of loiter immediately
-         */
+    // Corrected radius for altitude
+    const float loiter_radius = plane.nav_controller->loiter_radius(fabsf(plane.loiter.radius));
+    if (!is_positive(loiter_radius)) {
+        // Zero is invalid, protect against divide by zero for destination inside loiter radius case
         return true;
     }
 
-    // Bearing in centi-degrees
-    const int32_t bearing_cd = plane.current_loc.get_bearing_to(targetLoc);
-    return isHeadingLinedUp_cd(bearing_cd);
+    // Calculate relative position of the vehicle relative to loiter center projected onto the closest point of the loiter circle
+    // This removes error due to radial position as the nav controller attempts to track the circle
+    const Vector2f projected_pos = loiterCenterLoc.get_distance_NE(plane.current_loc).normalized() * loiter_radius;
```

Comment:
Note this is using floats, not `Vector2P` for double (if supported). Since we have a 10 deg theshold on the angle anyway I really don't think it will matter.